lib/ansible/plugins/inventory/tower.py

```diff
@@ -0,0 +1,201 @@
+from ansible.plugins.inventory import BaseInventoryPlugin
+
+try:
+    from urlparse import urljoin
+except ImportError:
+    from urllib.parse import urljoin
+
+try:
+    import requests
+    from requests.auth import HTTPBasicAuth
+except ImportError:
+    raise AnsibleError('You have to install python requests library to use tower inventory plugin.')
+
+
+class InventoryModule(BaseInventoryPlugin):
+    NAME = 'tower'
+    # Stays backward compatible with tower inventory script.
+    # If the user supplies 'tower_inventory' as path, the plugin will read from environment variables.
+    read_all_configs_from_env_var = False
+
+    def read_tower_inventory(self, tower_host, tower_user, tower_pass, inventory, license_type, verify_ssl=True):
+        if not re.match('(?:http|https)://', tower_host):
+            tower_host = "https://{tower_host}".format(tower_host=tower_host)
+        inventory_id = inventory.replace('/', '')
+        inventory_url = "/api/v2/inventories/{inv_id}/script/?hostvars=1&towervars=1&all=1".format(inv_id=inventory_id)
+        inventory_url = urljoin(tower_host, inventory_url)
+        config_url = urljoin(tower_host, "/api/v2/config/")
+        try:
+            if license_type != "open":
+                config_response = requests.get(config_url,
+                                               auth=HTTPBasicAuth(tower_user, tower_pass),
+                                               verify=verify_ssl)
+                if config_response.ok:
+                    source_type = config_response.json()['license_info']['license_type']
+                    if not source_type == license_type:
+                        raise AnsibleParserError("Tower server licenses must match: source: {source} local: {local}".
+                                                 format(source=source_type, local=license_type))
+                else:
+                    raise AnsibleParserError("Failed to validate the license of the remote Tower: {response}".
+                                             format(response=config_response))
+
+            response = requests.get(inventory_url,
+                                    auth=HTTPBasicAuth(tower_user, tower_pass),
+                                    verify=verify_ssl)
+            if not response.ok:
+                # If the GET /api/v2/inventories/N/script is not HTTP 200, print the error code
+                raise AnsibleParserError("Connection to remote host failed: {response}".format(response=response))
+            try:
+                # Attempt to parse JSON
+                return response.json()
+            except (ValueError, TypeError) as e:
+                # If the JSON parse fails, print the ValueError
+                raise AnsibleParserError("Failed to parse json from host: {err}".format(err=e))
+        except requests.ConnectionError as e:
+            raise AnsibleParserError("Connection to remote host failed: {err}".format(err=e))
```

Comment:
wrap exceptions in `to_native` when stringifying them

Reply:
👍 Will do!